File: src/db/controller.ts
import { getDB } from "./database";
import { randomUUID } from "expo-crypto";

// إضافة كتاب جديد
export async function addBook(bookData: any) {
    const db = getDB();
    const id = randomUUID();

    await db.execAsync(`
        INSERT INTO books (id, StudentName, BookName, BookCode, BorrowDate, ReturnDate)
        VALUES (
            '${id}',
            '${bookData.StudentName}',
            '${bookData.BookName}',
            '${bookData.BookCode ?? ""}',
            '${bookData.BorrowDate ? bookData.BorrowDate.toISOString() : ""}',
            '${bookData.ReturnDate ? bookData.ReturnDate.toISOString() : ""}'
        );
    `);

    return true;
}

// جلب الكتب مع الترتيب والبحث والصفحات
export async function getAllBooks({
    sort = "asc",
    searchText = "",
    page = 1,
    sortBy = "StudentName",
    size = 1000000000,
}) {
    const db = getDB();

    const offset = (page - 1) * size;
    const order = sort === "desc" ? "DESC" : "ASC";

    // WHERE
    let where = "";
    if (searchText.trim() !== "") {
        const term = `%${searchText}%`;
        where = `
            WHERE StudentName LIKE '${term}'
            OR BookName LIKE '${term}'
            OR BookCode LIKE '${term}'
        `;
    }

    // إجمالي النتائج
    const totalRow: any = await db.getFirstAsync(
        `SELECT COUNT(*) AS total FROM books ${where}`
    );
    const total = totalRow.total;

    // البيانات
    const data = await db.getAllAsync(
        `SELECT *
         FROM books
         ${where}
         ORDER BY ${sortBy} ${order}
         LIMIT ${size}
         OFFSET ${offset}`
    );

    return {
        data,
        total,
        page,
        size,
        pages: Math.ceil(total / size),
    };
}

// حذف كتاب
export async function deleteBook(id: string) {
    const db = getDB();

    await db.execAsync(`
        DELETE FROM books WHERE id = '${id}';
    `);

    return true;
}

// تحديث كتاب
export async function updateBook(id: string, updatedData: any) {
    const db = getDB();

    const fields: string[] = [];

    Object.keys(updatedData).forEach((key) => {
        let value = updatedData[key];

        // تحويل التاريخ في حال كان Date
        if (value instanceof Date) {
            value = value.toISOString();
        }

        // منع أخطاء الاقتباس داخل النص
        if (typeof value === "string") {
            value = value.replace(/'/g, "''");
        }

        fields.push(`${key}='${value}'`);
    });

    await db.execAsync(`
        UPDATE books 
        SET ${fields.join(", ")}
        WHERE id = '${id}';
    `);

    return true;
}

File: lib/db/controller.dart
import 'package:sqflite/sqflite.dart';
import 'package:uuid/uuid.dart';
import 'database.dart';

const _uuid = Uuid();

Future<bool> addBook(Map<String, dynamic> bookData) async {
  final db = getDB();
  final id = _uuid.v4();

  final row = {
    'id': id,
    'StudentName': bookData['StudentName'] ?? '',
    'BookName': bookData['BookName'] ?? '',
    'BookCode': bookData['BookCode'] ?? '',
    'BorrowDate': bookData['BorrowDate']?.toIso8601String() ?? '',
    'ReturnDate': bookData['ReturnDate']?.toIso8601String() ?? '',
  };

  await db.insert('books', row);
  return true;
}

Future<Map<String, dynamic>> getAllBooks({
  String sort = 'asc',
  String searchText = '',
  int page = 1,
  String sortBy = 'StudentName',
  int size = 1000000000,
}) async {
  final db = getDB();
  final offset = (page - 1) * size;
  final order = sort == 'desc' ? 'DESC' : 'ASC';

  String where = '';
  List<dynamic> whereArgs = [];
  if (searchText.trim().isNotEmpty) {
    final term = '%$searchText%';
    where = "WHERE StudentName LIKE ? OR BookName LIKE ? OR BookCode LIKE ?";
    whereArgs = [term, term, term];
  }

  final totalRes = await db.rawQuery(
    'SELECT COUNT(*) AS total FROM books $where',
    whereArgs,
  );
  final total = Sqflite.firstIntValue(totalRes) ?? 0;

  final data = await db.rawQuery(
    'SELECT * FROM books $where ORDER BY $sortBy $order LIMIT ? OFFSET ?',
    [...whereArgs, size, offset],
  );

  return {
    'data': data,
    'total': total,
    'page': page,
    'size': size,
    'pages': (total / size).ceil(),
  };
}

Future<bool> deleteBook(String id) async {
  final db = getDB();
  await db.delete('books', where: 'id = ?', whereArgs: [id]);
  return true;
}

Future<bool> updateBook(String id, Map<String, dynamic> updatedData) async {
  final db = getDB();
  final data = <String, dynamic>{};

  updatedData.forEach((key, value) {
    if (value is DateTime)
      data[key] = value.toIso8601String();
    else
      data[key] = value?.toString() ?? '';
  });

  await db.update('books', data, where: 'id = ?', whereArgs: [id]);
  return true;
}
